polynote-spark/src/test/scala/polynote/kernel/remote/RemoteSparkKernelSpec.scala

```diff
@@ -0,0 +1,182 @@
+class RemoteSparkKernelSpec extends FreeSpec with Matchers {
+
+  private object testMonitor
+
+  private implicit val executionContext: ExecutionContext = ExecutionContext.fromExecutorService(
+    Executors.newCachedThreadPool(new ThreadFactory {
+      def newThread(r: Runnable): Thread = {
+        val thread = new Thread(r)
+        thread.setDaemon(true)
+        thread.setName("Test runner thread")
+        thread
+      }
+    })
+  )
+
+  private implicit val contextShift: ContextShift[IO] = IO.contextShift(executionContext)
+  private implicit val timer: Timer[IO] = IO.timer(executionContext)
+
+  private def runTest(fn: => IO[Unit])(implicit pos: Position): Unit = testMonitor.synchronized {
+    contextShift.evalOn(executionContext)(fn).unsafeRunSync()
+  }
+
+  private val path = ShortString("testnotebook")
+
+  private val outputMessages = Queue.unbounded[IO, Message].unsafeRunSync()
+  private val statusUpdates = outputMessages.contramap[KernelStatusUpdate](KernelStatus(path, _))
+
+  private val initialNotebook = Notebook(
+    path,
+    ShortList(List(
+      NotebookCell(0.toShort, "scala", "val foo = 22")
+    )),
+    None
+  )
+
+  private val config = PolynoteConfig()
+  private def mockKernel = new MockKernel(initialNotebook)
+  private def mockKernelFactory = new MockKernelFactory(mockKernel)
+
+  "protocol handshake" in {
+    runTest {
+      val transport = new LocalTestTransport
+      val currentNotebook = Ref[IO].of(initialNotebook).unsafeRunSync()
+
+      transport.expect { _
+        .response {
+          case Announce(_) => true
+        }
+        .request {
+          case InitialNotebook(_, `initialNotebook`) => true
+        }
+        .response {
+          case UnitResponse(_) => true
+        }
+      }
+
+      transport.run {
+        await =>
+          for {
+            startKernelF <- RemoteSparkKernel(
+              statusUpdates,
+              () => currentNotebook.get,
+              config,
+              transport).start
+            remoteClient = new RemoteSparkKernelClient(transport.client, mockKernelFactory)
+            runRemote <- remoteClient.run().start
+            remoteKernel <- startKernelF.join
+            _ <- await
+            _ <- remoteKernel.shutdown()
+            _ <- runRemote.join
+          } yield ()
+      }
+    }
+  }
+
+  "run cell, handle mapping, stream proxy" in {
+    runTest {
+      val transport = new LocalTestTransport
+      for {
+        currentNotebook   <- Ref[IO].of(initialNotebook)
+        startRemoteKernel <- RemoteSparkKernel(
+          statusUpdates,
+          () => currentNotebook.get,
+          config,
+          transport).start
+
+        remoteClient = new RemoteSparkKernelClient(transport.client, mockKernelFactory)
+        runRemoteClient <- remoteClient.run().start
+        remoteKernel <- startRemoteKernel.join
+        _ <- remoteKernel.init
+
+        results <- remoteKernel.runCell(2.toShort).flatMap(_.compile.toList)
```

Comment:
Took me a while to figure out where cell 2 came from since it wasn't in `initialNotebook`. I wonder whether it might be possible to refactor such that all the mock cell definitions are in one place right next to each other? Or is there a reason for them to be separated as they are now?

Reply:
Agreed it's pretty lame that the mock kernel basically ignores the notebook. I guess we should just give it a `Notebook` that we use instead of creating a dummy one in the test